src/demo-app/data-table/person-data-source.ts

```diff
@@ -1,29 +1,58 @@
-import {CollectionViewer, DataSource} from '@angular/material';
+import {CollectionViewer, DataSource, MdPaginator, PageChangeEvent} from '@angular/material';
 import {Observable} from 'rxjs/Observable';
 import {PeopleDatabase, UserData} from './people-database';
+import {BehaviorSubject} from 'rxjs/BehaviorSubject';
 
 export class PersonDataSource extends DataSource<any> {
+  /** Data that should be displayed by the table. */
+  _displayData = new BehaviorSubject<UserData[]>([]);
+
+  /** Cached data provided by the display data. */
   _renderedData: any[] = [];
 
-  constructor(private _peopleDatabase: PeopleDatabase) {
+  constructor(private _peopleDatabase: PeopleDatabase,
+              private _paginator: MdPaginator) {
     super();
+
+    // Subscribe to page changes by clearing the cached data and
+    // determining the updated display data.
+    this._paginator.pageChange.subscribe(() => {
+      this._renderedData = [];
+      this.updateDisplayData();
+    });
   }
 
   connect(collectionViewer: CollectionViewer): Observable<UserData[]> {
-    return collectionViewer.viewChange.map((view: {start: number, end: number}) => {
+    this.updateDisplayData();
+
+    const streams = [collectionViewer.viewChange, this._displayData];
+    return Observable.combineLatest(streams).map((results: any[]) => {
+      const view: {start: number, end: number} = results[0];
```

Comment:
The downside is that I lose the typing of view, unless there's a syntax I'm missing that would let me apply different typing to view and data